### PR TITLE
Chat

### PR DESCRIPTION
**Pull request by the-itz**

### CODE REVIEW

File: src/pages/rooms/chat-input-form.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  //onEdit: (input: string, callback: Dispatch<SetStateAction<string>>) => void;
+  onCancel: (input: string) => void;
+  preSetInput: string;
+  setInput: Dispatch<SetStateAction<string>>;
+  users?: User[];
+  allowEscape: boolean
+}
+
+export default function ChatInputForm({ onCreate, onCancel, setInput, preSetInput, users, allowEscape }: Props) {
+  const [userInput, setUserInput] = useState(preSetInput);
```
Comment (arminfro) [inline src/pages/rooms/chat-input-form.tsx:22]: I'd recommend to rename `userInput` to just `input`, `setUserInput` to `setInput`

File: src/pages/rooms/chat-input-form.tsx [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    } else if (allowEscape && e.key == 'Escape') {
+      onCancel(oldMsg);
+    } else if (e.key == '@') {
+      setOldMsg(userInput)
+      setDoAtmention(true)
+    }
+  };
+
+  const selectAtMention = (index: number, value: string): void => {
+
```
Comment (arminfro) [inline src/pages/rooms/chat-input-form.tsx]: this new line is not needed

File: src/pages/rooms/chat-input-form.tsx [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    setUserInput(oldMsg + '**@' + value + '** ');
```
Comment (arminfro) [inline src/pages/rooms/chat-input-form.tsx]:
```suggestion
    setUserInput(oldMsg + '**@' + value + '** ');
```
```suggestion
    setUserInput(oldMsg + ' **@' + value + '** ');
```
Having a space before `**@...` makes sense, I guess.
Edit: Maybe we should check if there is a space before, if not insert one

File: src/pages/rooms/chat-input-form.tsx [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  return (
+    <form className="ui reply form" >
+      <div className="dropdown-container">
+        <textarea
+          value={userInput}
+          onChange={onChange}
+          onKeyDown={onKeyDown}
+          placeholder="Your Message"
+          id="msg"
+        ></textarea>
```
Comment (arminfro) [inline src/pages/rooms/chat-input-form.tsx]: this element has no childs, can be closed immediately

File: src/pages/rooms/chat-input-form.tsx [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        {doAtMention ? <Dropdown entries={reducedUserList.map((u) => u.name)} callback={selectAtMention} action={action} /> : ''}
```
Comment (arminfro) [inline src/pages/rooms/chat-input-form.tsx]: Logical `&&` is sufficient, no ternary operator needed

File: src/pages/rooms/chat-input-form.tsx [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+      </div>
+      <button onClick={onClickSubmit} className="ui blue labled submit icon button ">
+        <i className="icon edit"></i>send
+      </button>
+      <span className="actions">
+        or <b>Ctrl-Return</b>
+      </span>
+
+      {allowEscape && <button onClick={onClickCancel} className="ui red labled submit icon button ">
```
Comment (arminfro) [inline src/pages/rooms/chat-input-form.tsx]: `<button ...` should go on the next line, to keep indentation level of opening and closing tag

File: src/pages/rooms/chat-input-form.tsx [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        <i className="icon cancel"></i>cancel
```
Comment (arminfro) [inline src/pages/rooms/chat-input-form.tsx]: indentation is wrong

File: src/pages/rooms/chat-input-form.tsx [hunk 8]
```diff
@@ -0,0 +1,130 @@
+import React, {
```
Comment (arminfro) [inline src/pages/rooms/chat-input-form.tsx]: Many functions in this file have no return type specified

File: src/pages/rooms/chat-message.tsx [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    setEdit(false);
+  };
+
+  const onCancel = (msg: string) => {
+    resetStatus();
+  }
+
+  const createReplyString = (chat: Chat, newInput: string): string => {
+    return `Reply to ${chat.user.name}'s Message on ${new Date(chat.updatedAt).toLocaleDateString()}: \n ${chat.msg.replace(/^/gm, "> ")} \n\n  ${newInput}`
+  }
```
Comment (arminfro) [inline src/pages/rooms/chat-message.tsx]: missing new line below

File: src/pages/rooms/chat-message.tsx [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+          <span className="date">
+            {formatDistance(new Date(chat.updatedAt), new Date(), {
+              includeSeconds: true,
+            })}
+          </span>
+          <span className="date">
+            {new Date(chat.updatedAt).getTime() !==
+              new Date(chat.createdAt).getTime()
+              ? '(edited)'
+              : ''}
```
Comment (arminfro) [inline src/pages/rooms/chat-message.tsx]: Logical `&&` is sufficient, no ternary operator needed

File: src/pages/rooms/chat-message.tsx [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+              <a className="avatar" href={l.url}>
+                <img src={l.imgUrl} width="100" height="100" />
+              </a>
+              <a className="metadata" href={l.url}>
+                <p><b>{l.title}</b></p> <br />
+                {l.description}
+              </a>
+            </div>
+          )
+        ) : ('')
```
Comment (arminfro) [inline src/pages/rooms/chat-message.tsx]: Logical `&&` is sufficient, no ternary operator needed

File: src/pages/rooms/chat-message.tsx [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+            })}
+          </span>
+          <span className="date">
+            {new Date(chat.updatedAt).getTime() !==
+              new Date(chat.createdAt).getTime()
+              ? '(edited)'
+              : ''}
+          </span>
+        </div>
+        {(chat.links) ? (
```
Comment (arminfro) [inline src/pages/rooms/chat-message.tsx]: return statement is quite long, maybe move this code into a new `ChatLink` component

File: src/pages/rooms/chats.tsx [hunk 13]
```diff
@@ -27,30 +37,49 @@ export default function Chats({ roomId }: Props): ReactElement {
...
+
+  const onReply = (chat: Chat) => {
+    setInput(chat.msg)
+  }
+
   return (
-    <div>
-      <div className="ui relaxed divided list">
+    <div className="ui container" style={{ maxHeight: '900px' }}>
+      <br></br>
```
Comment (arminfro) [inline src/pages/rooms/chats.tsx]: this element has no childs, can be closed immediately

File: src/pages/rooms/dropdown.tsx [hunk 14]
```diff
@@ -0,0 +1,53 @@
+import React, { useEffect, useLayoutEffect, useRef, useState } from 'react';
+
+export interface Position {
```
Comment (arminfro) [inline src/pages/rooms/dropdown.tsx:3]: is not used

File: src/pages/rooms/dropdown.tsx [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+interface Props {
+  entries: string[];
+  callback: (index: number, value: string) => void
+  action: string
+}
+
+
+export default function Dropdown({ entries, callback, action }: Props) {
```
Comment (arminfro) [inline src/pages/rooms/dropdown.tsx]: return type is missing

File: src/pages/rooms/dropdown.tsx [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+  const [selected, setSelected] = useState(0);
+  const dropdownRef = useRef(null);
+
+  useEffect(() => {
+    onAction(action)
+    return
```
Comment (arminfro) [inline src/pages/rooms/dropdown.tsx]: `return` is not needed

File: src/pages/rooms/dropdown.tsx [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+    } else if (action == 'ArrowDown') {
+      setSelected((selected + 1) % entries.length)
+    } else if (action == 'ArrowUp') {
+      setSelected((Math.abs(selected - 1)) % entries.length)
+    }
+  };
+
+  return (
+    <div className="dropdown" >
+      <input type="hidden" name="input"></input>
```
Comment (arminfro) [inline src/pages/rooms/dropdown.tsx]: hidden input is needed for what?

File: src/pages/rooms/dropdown.tsx [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+      <div className="dropdowntext" ref={dropdownRef}>
+        {entries.map((e, i) => {
+          return (
+            <div key={i} data-value={i} className={(selected == i) ? "selected" : ""} onClick={(ev) => select(i, e)}>
```
Comment (arminfro) [inline src/pages/rooms/dropdown.tsx]: `data-value` is not needed, I think
Logical `&&` is sufficient, no ternary operator needed

**Review by arminfro — CHANGES_REQUESTED** [review]

File: src/pages/rooms/chat-input-form.tsx (hunk 1 above)
Comment (the-itz) [reply src/pages/rooms/chat-input-form.tsx:22]: Not sure about this. I have userInput for the what the user inputs and
preSetInput the text which the input form should display when there ist already something to edit
and setInput is the callback which is used  to trigger saving to the db + rerender.
The latter one would clash with your proposal. 
If you want to have the change. I would then rename the callback to setInputClbk

File: src/pages/rooms/chat-input-form.tsx (hunk 2 above)
Comment (the-itz) [reply src/pages/rooms/chat-input-form.tsx]: Done

File: src/pages/rooms/chat-input-form.tsx (hunk 3 above)
Comment (the-itz) [reply src/pages/rooms/chat-input-form.tsx]: done, incl check
Found a new issue there. We have to define the behaviour if the user presses @ in the middle of the text. Currently we just add the string to the end of the message. I'll create an issue for this.

File: src/pages/rooms/chat-input-form.tsx (hunk 4 above)
Comment (the-itz) [reply src/pages/rooms/chat-input-form.tsx]: done

File: src/pages/rooms/chat-input-form.tsx (hunk 5 above)
Comment (the-itz) [reply src/pages/rooms/chat-input-form.tsx]: done

File: src/pages/rooms/chat-input-form.tsx (hunk 3 above)
Comment (the-itz) [reply src/pages/rooms/chat-input-form.tsx]: is used the ternary operator here, didn't work with && though
Also tried to have ist inline in the setUserInput call -> messed up result. So preset const string used this to assemble result string.

File: src/pages/rooms/chat-input-form.tsx (hunk 6 above)
Comment (the-itz) [reply src/pages/rooms/chat-input-form.tsx]: done

File: src/pages/rooms/chat-input-form.tsx (hunk 7 above)
Comment (the-itz) [reply src/pages/rooms/chat-input-form.tsx]: Don't get this one, could you pls check the update of the file?

File: src/pages/rooms/chat-input-form.tsx (hunk 8 above)
Comment (the-itz) [reply src/pages/rooms/chat-input-form.tsx]: Done. Defined Return types of functions.

File: src/pages/rooms/chat-message.tsx (hunk 9 above)
Comment (the-itz) [reply src/pages/rooms/chat-message.tsx]: If you meant below  / after function, then done

File: src/pages/rooms/chat-message.tsx (hunk 10 above)
Comment (the-itz) [reply src/pages/rooms/chat-message.tsx]: Ok, done

File: src/pages/rooms/chat-message.tsx (hunk 11 above)
Comment (the-itz) [reply src/pages/rooms/chat-message.tsx]: Done

File: src/pages/rooms/chats.tsx (hunk 13 above)
Comment (the-itz) [reply src/pages/rooms/chats.tsx]: Done

File: src/pages/rooms/dropdown.tsx (hunk 14 above)
Comment (the-itz) [reply src/pages/rooms/dropdown.tsx:3]: Done

File: src/pages/rooms/dropdown.tsx (hunk 15 above)
Comment (the-itz) [reply src/pages/rooms/dropdown.tsx]: Done

File: src/pages/rooms/dropdown.tsx (hunk 16 above)
Comment (the-itz) [reply src/pages/rooms/dropdown.tsx]: Done

File: src/pages/rooms/dropdown.tsx (hunk 17 above)
Comment (the-itz) [reply src/pages/rooms/dropdown.tsx]: Removed. Done.

File: src/pages/rooms/dropdown.tsx (hunk 18 above)
Comment (the-itz) [reply src/pages/rooms/dropdown.tsx]: Jepp not needed, removed and Done

File: src/pages/rooms/chat-input-form.tsx (hunk 1 above)
Comment (the-itz) [reply src/pages/rooms/chat-input-form.tsx:22]: The thumb up means, you want me to change names? @arminfro

File: src/pages/rooms/chat-input-form.tsx (hunk 3 above)
Comment (the-itz) [reply src/pages/rooms/chat-input-form.tsx:94]: Ok, learned something new. If you create an issue out of a comment, the comment will then disappear.
Done and issues created.

File: src/pages/rooms/chat-input-form.tsx (hunk 1 above)
Comment (arminfro) [reply src/pages/rooms/chat-input-form.tsx:22]: it means you're right, leave it like it is. I didn't saw the prop with the name `setInput`

File: src/types/chat.ts [hunk 19]
```diff
@@ -17,6 +18,7 @@ export default interface Chat {
   readonly msg: string;
   readonly room: Room;
   readonly user: User;
+  readonly links: Link[];
```
Comment (the-itz) [inline src/types/chat.ts]: @arminfro after the pull I've got this line twice. I'll remove the second one on my branch.

**Pull request merged**